// @superviz/threejs-plugin/src/components/AvatarName.ts
import { Object3D } from 'three';
import * as THREE from 'three';
import SpriteText from 'three-spritetext';

import { Colors, TextColors } from '../common/types/colors.types';

export class AvatarName {
  constructor(name: string, slot: number, height: number, scale: number) {
    this.name = name;
    this.scale = scale;
    if (slot) {
      this.color = Colors[slot];
    } else {
      const firstColor = Colors[0];
      this.color = firstColor;
    }
    this.height = height;
    this.slot = slot;
  }

  public model: Object3D;
  private name: string;
  private color: string;
  private scale: number;
  private height: number;
  public slot: number;
  public text: SpriteText;

  load = async () => {
    const y = this.height;
    const font = new FontFace(
      'OpenSans-SemiBold',
      'url(https://superviz2homologmediaserver.s3.amazonaws.com/static/fonts/OpenSans-SemiBold.woff)',
    );
    await font.load();

    document.fonts.add(font);
    const texGeo = this.createTextGeometry(this.name, this.color);
    this.text = texGeo;
    const root = new Object3D();
    root.add(texGeo);
    texGeo.position.set(0, y, 0);
    this.model = root;
    return this.model;
  };

  createTextGeometry = (text: string, color: string) => {
    const { scale } = this;
    // const scale = 0.005;
    const textColor = TextColors[this.slot];
    const backgroundColor = color;
    const myText = new SpriteText(text, 0.1);
    myText.color = textColor;
    myText.padding = 2.7;
    myText.backgroundColor = backgroundColor;
    myText.fontFace = 'OpenSans-SemiBold';
    myText.fontSize = 100;
    myText.material.sizeAttenuation = true;
    myText.borderRadius = 8;

    myText.scale.set(myText.scale.x * scale, myText.scale.y * scale, myText.scale.z * scale);
    return myText;
  };
}
